File: l17.py
""" Letter Combinations of a Phone Number """

class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        ans = []
        cdic = {1:[]}
        a = 'abcdefghijklmnopqrstuvwxyz'
        j = 0
        for i in range(2,7):
            if j < len(a):
                cdic[i] = [a[j],a[j+1],a[j+2]]
                j += 3
        
        cdic[7] = ['p','q','r','s']
        cdic[8] = ['t','u','v']
        cdic[9] = ['w','x','y','z']

        def dfs(snum, pr):
            if not snum:
                ans.append(pr)
                return
            for i in cdic[int(snum[0])]:
                dfs(snum[1:],pr+i)
        
        dfs(digits,'')
        return ans




            
s = Solution()
print(s.letterCombinations('23'))

File: l17.java
import java.util.ArrayList;

// Letter Combinations of a Phone Number

class Solution {
    public List<String> letterCombinations(String digits) {
        String[] map = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<String> ans = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return ans;
        }
        helper(map, ans, digits, 0, "");
        return ans;
    }

    private void helper(String[] map, List<String> ans, String digits, int ind, String cur) {
        if (ind == digits.length()) {
            ans.add(cur);
            return;
        }
        String possible = map[Integer.parseInt(digits.substring(ind,ind + 1))];
        for (int i=0; i < possible.length(); i++) {
            helper(map, ans, digits, ind+1, cur + possible.charAt(i));
        }
        return;
    }
}
